WIP: OIDC integration

mdp opened the pull request:
Just pushing this up for a quick sanity check.

mdp committed:
WIP: OIDC Auth

mdp committed:
Add a map to keyword test

princemaple commented on apps/fz_common/lib/fz_map.ex:
```diff
@@ -28,4 +28,24 @@ defmodule FzCommon.FzMap do
+        try do
+          String.to_existing_atom(k)
+        rescue
+          ArgumentError -> String.to_atom(k)
```
This is the same as simply `String.to_atom` without `try`, isn't it?
Not that it's a recommendation. I'll add my recommendation to the config line.

mdp replied:
Yep, I misunderstood the docs and now realize that String.to_atom('foo') run multiple times will result in the same atom behind the scenes, similar to a Ruby symbol.

princemaple commented on config/runtime.exs:
```diff
@@ -239,3 +239,19 @@ if config_env() == :prod do
+    |> Enum.map(fn provider ->
+      {name, value} = provider
+      {name, put_in(value, [:redirect_uri], "#{external_url}/auth/oidc/#{name}/callback/")}
+    end)
```
I think it's better to `Jason.decode(_, keys: :atoms)` then work with the maps directly.
You probably don't need to change your code other than removing the `map_to_keyword_list` call.

mdp replied:
Thanks, yeah I didn't know that was an option. I should have dived more into the Jason library.

princemaple commented on apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex:
```diff
@@ -48,6 +49,29 @@ defmodule FzHttpWeb.AuthController do
+    %{"provider" => provider} = params
+
+    with {:ok, tokens} <- OpenIDConnect.fetch_tokens(String.to_atom(provider), params),
+         {:ok, claims} <- OpenIDConnect.verify(String.to_atom(provider), tokens["id_token"]) do
```
Probably should have a list of providers somewhere and always use `to_existing_atom`

Or even build a multi clause function with some macro to do the conversion directly.

jamilbk replied:
@princemaple The challenge is the list of providers will be configurable by the user at runtime. Ultimately it'll be defined in `/etc/firezone/firezone.rb` -- I think we're hoping to address #501 using a more general approach here so that we can support any OIDC-based provider.

`String.to_atom/1` is likely [problematic here though](http://www.moxleystratton.com/when-to-use-elixirs-string-to-atom/) because `provider` is end-user-supplied, right? Hmm. We may have to hack the OpenIDConnect module to support loading config from maps vs keyword lists?

princemaple reviewed (commented):
Hi, not sure if that's the sanity check you wanted.. left some comments

jamilbk commented on omnibus/cookbooks/firezone/attributes/default.rb:
```diff
@@ -99,13 +99,31 @@
+    client_secret: "CLIENT_SECRET",
+    redirect_uri: "https://example.com/session",
+    response_type: "code",
+    scope: "openid email profile"
```
Missing comma at EOL

mdp replied:
Thanks, Fixed in 0fbcb2e2

mdp committed:
Merge up with master

mdp committed:
Remove atom creation, only atomize config fields

mdp committed:
Fixing missing trailing comma https://github.com/firezone/firezone/pull/573/files#r865472611

mdp committed:
Remove existing atom call

mdp committed:
Conditionally run openid_connect, remove debug Logger

mdp committed:
Clean up and comments

princemaple commented on config/runtime.exs:
```diff
@@ -239,3 +240,27 @@ if config_env() == :prod do
+          |> Map.put("redirect_uri", "#{external_url}/auth/oidc/#{provider}/callback/")
+        )
+      )
+    end)
```
I think we can simplify this to a simple pipeline and remove the custom map to keyword list function. It's a rather dangerous function and we don't want people to use it too freely. Plus, as the structure of the settings is flat (correct me if I'm wrong) I don't think we need to recursively parse it anymore.

```elixir
|> Map.new(fn {provider, settings} ->
  {provider,
   settings
   |> Map.take(
     ~w(discovery_document_uri client_id client_secret redirect_uri response_type scope label)
   )
   |> Map.put("redirect_uri", "#{external_url}/auth/oidc/#{provider}/callback/")
   |> Enum.map(fn {k, v} -> {String.to_existing_atom(k), v} end)}
end)
```

We can use `to_existing_atom` here because they are defined in `:openid_connect`.

mdp replied:
Thanks @princemaple , this is much cleaner, and yes, the structure of the map is flat so we can get rid of the custom helper. I'll make the changes tomorrow morning and do some quick testing.

mdp replied:
So openid_connect does create these atoms, but, I believe, not until the Worker is created with the passed in configuration. So existing_atom fails in this instance. I've changed it to `to_atom`, which should be safe because of the previous pipeline `take` which limits the number of potential atoms. @princemaple thoughts?

jamilbk replied:
@mdp I know it's a little more verbose, but how about more explicitly mapping the atoms in a Keyword list like this?

```elixir
Map.new(fn {provider, settings} ->
  {provider, [
    discovery_document_uri: settings["discovery_document_uri"],
    client_id: settings["client_id"],
    client_secret: settings["client_secret"],
    redirect_uri: "#{external_url}/auth/oidc/#{provider}/callback/",
    response_type: settings["response_type"],
    scope: settings["scope"],
    label: settings["label"]
  ]
end)
```

princemaple replied:
> So openid_connect does create these atoms, but, I believe, not until the Worker is created with the passed in configuration. So existing_atom fails in this instance.

Not until the worker is started, but you are right that `runtime.exs` is loaded before the libs are loaded and started. I'm happy with the `to_atom` here as it's behind a `take` and load time only. The verbose way also looks fine. I'll leave the decision to you guys.

mdp replied:
@jamilbk Yeah, I like this better, it's clearer about exactly what's happening.

princemaple reviewed (commented):
What do you think?

mdp committed:
Cleanup and test fixes

- Cleanup map to keyword list per suggestions from @princemaple
- Fix issues with openid_connect_providers not being set, default to
empty list

mdp committed:
Fix existing atom issue

openid_connect creates these atoms when it's Worker is created.
Therefore the atoms don't yet exist when we are creating the config.
Since we are using 'take' in the pipeline, the creation of these atoms should
be fairly safe.

jamilbk commented on apps/fz_http/lib/fz_http_web/controllers/auth_controller.ex:
```diff
@@ -48,6 +49,29 @@ defmodule FzHttpWeb.AuthController do
+          |> request(%{})
+      end
+    else
+      _ -> send_resp(conn, 401, "")
```
@mdp Would it make sense to log the error message here with something like (apologies have not tested this):

```
{:error, reason} ->
  Logger.warn("OpenIDConnect Error: Could not fetch tokens from params. Reason: #{reason}")
  send_resp(conn, 401, "")
```

Just so the administrator has something to debug if this fails?

mdp replied:
Yeah, I think that makes sense.

jamilbk commented on omnibus/cookbooks/firezone/attributes/default.rb:
```diff
@@ -99,13 +99,31 @@
+}
+
+# DEPRECATED
+# Previously, Firezone used preconfigured Oauth2 providers. We've moved to OIDC authentication
```
🎉

mdp committed:
Mock openid wip

mdp committed:
Mocking and more tests

mdp committed:
Clearer mapping to OIDC config and add logs

- More verbose but clearer mapping of settings to OpenId Connect
provider configs
- Add a Logger.warn for errors on OIDC to improve debugging for admins

mdp committed:
Add OIDC configuration docs

mdp committed:
Merge up with master

mdp committed:
Match config to openid_connect expectations

jamilbk commented on omnibus/cookbooks/firezone/attributes/default.rb:
```diff
@@ -99,13 +99,31 @@
+    discovery_document_uri: "https://accounts.google.com/.well-known/openid-configuration",
+    client_id: "CLIENT_ID",
+    client_secret: "CLIENT_SECRET",
+    redirect_uri: "https://example.com/session",
```
Would it help to use a more realistic example redirect_uri here? e.g.
 `"https://firezone.example.com/auth/oidc/google/callback"`

mdp replied:
Fixed in https://github.com/firezone/firezone/pull/573/commits/948b65d4b4fbf760b7f0dd5a57aa10d20c7566fe

jamilbk commented on .oidc_env.sample.json:
```diff
@@ -0,0 +1,11 @@
+    "discovery_document_uri": "https://accounts.google.com/.well-known/openid-configuration",
+    "client_id": "CLIENT_ID",
+    "client_secret": "CLIENT_SECRET",
+    "redirect_uri": "https://example.com/session",
```
Would it be more clear to use something like `https://firezone.example.com/auth/oidc/google/callback` so the user is aware of the path format requirement?

jamilbk reviewed (commented):
LGTM, only the comment Re: example `redirect_uri`

@mdp you should have `write` privs on this repo -- did you want to push up a feature branch to get the CI to kick off? If not no worries, I can merge to master and test the whole enchilada after :-)

mdp committed:
Better example of redirect_uri
